File: examples/erlang/05-data-structures/data_structures.erl
%% Data Structures in Erlang
%% =========================
%% Erlang provides lists, tuples, maps, and records.
%% Immutable data structures designed for concurrent, fault-tolerant systems.

-module(data_structures).
-export([main/0]).

%% ── Lists (linked lists) ─────────────────────────────

demonstrate_lists() ->
  io:format("--- Lists (linked, immutable) ---~n"),

  %% List literal (syntactic sugar for cons cells)
  Fruits = [apple, banana, cherry],
  io:format("fruits = ~w~n", [Fruits]),

  %% Head and tail (pattern matching)
  [First | Rest] = Fruits,
  io:format("first = ~w~n", [First]),
  io:format("rest = ~w~n", [Rest]),

  %% Indexing (inefficient O(n))
  io:format("fruits[0] = ~w~n", [lists:nth(0, Fruits)]),
  io:format("fruits[2] = ~w~n", [lists:nth(2, Fruits)]),

  %% cons (prepend)
  Fruits2 = [date | Fruits],
  io:format("After cons date: ~w~n", [Fruits2]),

  %% length
  io:format("length = ~p~n", [length(Fruits)]),

  %% List comprehensions
  Squares = [X * X || X <- lists:seq(1, 10)],
  io:format("squares (1-10): ~w~n", [Squares]),

  Evens = [X || X <- lists:seq(1, 20), X rem 2 =:= 0],
  io:format("evens (1-20): ~w~n", [Evens]).

%% ── Tuples (fixed-size, heterogenous) ─────────────

demonstrate_tuples() ->
  io:format("~n--- Tuples (fixed-size) ---~n"),

  %% Tuple literal
  Point = {10, 20},
  io:format("point = ~w~n", [Point]),

  %% Element access by index
  io:format("point[0] = ~p~n", [element(1, Point)]),
  io:format("point[1] = ~p~n", [element(2, Point)]),

  %% Pattern matching (destructuring)
  {X, Y} = Point,
  io:format("X = ~p, Y = ~p~n", [X, [Y]),

  %% Tuple with many elements
  RGB = {255, 0, 128},
  io:format("RGB = ~w~n", [RGB]).

%% ── Maps (key-value) ─────────────────────────────

demonstrate_maps() ->
  io:format("~n--- Maps (key-value) ---~n"),

  %% Map literal
  Person = #{name => "Ada", age => 36, city => "London"},
  io:format("person = ~w~n", [Person]),

  %% Access
  io:format("name = ~s~n", [maps:get(name, Person)]),
  io:format("age = ~p~n", [maps:get(age, Person)]),

  %% Check existence
  io:format("has email? ~p~n", [maps:is_key(email, Person)]),

  %% Update (add or modify)
  Person2 = maps:put(email, "ada@example.com", Person),
  io:format("After add email: ~w~n", [Person2]),

  Person3 = maps:remove(city, Person2),
  io:format("After remove city: ~w~n", [Person3]),

  %% Get all keys and values
  io:format("keys: ~w~n", [maps:keys(Person)]),
  io:format("values: ~w~n", [maps:values(Person)]),

  %% Fold (reduce)
  Sum = lists:foldl(fun(X, Acc) -> X + Acc end, 0, [1, 2, 3, 4, 5]),
  io:format("fold sum = ~p~n", [Sum]),

  %% Map (transform)
  Doubled = maps:map(fun(_K, V) -> V * 2 end, #{a => 1, b => 2, c => 3}),
  io:format("doubled map: ~w~n", [Doubled]).

%% ── Sets ───────────────────────────────────────────

demonstrate_sets() ->
  io:format("~n--- Sets (unique elements) ---~n"),

  %% Set literal (duplicates removed)
  Tags = sets:from_list([python, typing, python, data]),
  io:format("tags = ~w~n", [Tags]),
  io:format("size = ~p~n", [sets:size(Tags)]),

  %% Add element
  Tags2 = sets:add_element(algorithms, Tags),
  io:format("After add 'algorithms': ~w~n", [Tags2]),

  %% Remove element
  Tags3 = sets:del_element(data, Tags2),
  io:format("After remove 'data': ~w~n", [Tags3]),

  %% Check membership
  io:format("is_member 'typing'? ~p~n", [sets:is_element(typing, Tags)]),

  %% Set operations
  SetA = sets:from_list([1, 2, 3, 4, 5]),
  SetB = sets:from_list([4, 5, 6, 7, 8]),
  Union = sets:union(SetA, SetB),
  Intersection = sets:intersection(SetA, SetB),
  Difference = sets:subtract(SetA, SetB),

  io:format("union: ~w~n", [Union]),
  io:format("intersection: ~w~n", [Intersection]),
  io:format("difference A-B: ~w~n", [Difference]),

  %% Subset/superset
  io:format("subset? {2,3,4} of {1,2,3,4,5,6}? ~p~n",
    [sets:is_subset(sets:from_list([2, 3, 4]), SetA)]),
  io:format("superset? {1,2,3,4,5,6} of {2,3,4}? ~p~n",
    [sets:is_superset(SetA, sets:from_list([2, 3, 4]))]).

%% ── Records (named tuples) ─────────────────────────

-record(point, {x, y}).
-record(person, {name, age}).

demonstrate_records() ->
  io:format("~n--- Records (named tuples) ---~n"),

  %% Record creation
  P = #point{x = 10, y = 20},
  io:format("Point = ~w~n", [P]),

  %% Access fields (use #record.field)
  io:format("x = ~p~n", [P#point.x]),
  io:format("y = ~p~n", [P#point.y]),

  %% Pattern matching records
  #point{x = X, y = Y} = P,
  io:format("X = ~p, Y = ~p~n", [X, [Y]),

  %% Update records
  P2 = P#point{x = 30},
  io:format("After update x: ~w~n", [P2]),
  io:format("Original unchanged: ~w~n", [P]),

  %% Person record with multiple fields
  Ada = #person{name = "Ada", age = 36},
  io:format("Person = ~w~n", [Ada]),
  io:format("name = ~s~n", [Ada#person.name]),
  io:format("age = ~p~n", [Ada#person.age]).

%% ── Binary (binary data) ─────────────────────────

demonstrate_binary() ->
  io:format("~n--- Binaries (binary data) ---~n"),

  %% << for building binaries
  Byte1 = 16#FF,
  Byte2 = 16#AA,

  %% Concatenate binaries
  Combined = <<Byte1, Byte2>>,
  io:format("Combined = ~w~n", [Combined]),

  %% Size in bytes
  io:format("size = ~p~n", [byte_size(Combined)]),

  %% Extract segment
  <<First:8, _Last:8>> = Combined,
  io:format("First 8 bits = ~w~n", [First]),
  io:format("Last 8 bits = ~w~n", [_Last]),

  %% Bit syntax
  <<Bit1:1, _:7>> = Combined,
  io:format("First bit = ~p~n", [Bit1]).

%% ── Processes (lightweight actors) ─────────────────

demonstrate_processes() ->
  io:format("~n--- Processes (actors) ---~n"),

  %% Spawn process
  Pid = spawn(fun() ->
    receive
      hello -> io:format("Received: hello~n")
    end
  end),

  %% Send message
  Pid ! hello,

  %% Self-pid
  io:format("Process ID: ~w~n", [self()]),
  io:format("Spawned PID: ~w~n", [Pid]).

%% ── ETS (Erlang Term Storage) ─────────────

demonstrate_ets() ->
  io:format("~n--- ETS (Term Storage) ---~n"),

  %% Create named ETS table
  TableId = ets:new(counter_table, [set, named_table, public]),

  %% Insert
  ets:insert(TableId, {counter, 0}),

  %% Lookup
  case ets:lookup(TableId, counter) of
    [{counter, Value}] -> io:format("counter value = ~p~n", [Value]);
    [] -> io:format("counter not found~n")
  end,

  %% Update
  ets:update_counter(TableId, counter, {1}),
  io:format("After increment: ~p~n", [ets:lookup_element(2, TableId, counter)]),

  %% Delete table
  ets:delete(TableId).

%% ── Summary ───────────────────────────────────────────

main(_) ->
  io:format("Data Structures in Erlang~n"),
  io:format("========================~n~n"),

  demonstrate_lists(),
  demonstrate_tuples(),
  demonstrate_maps(),
  demonstrate_sets(),
  demonstrate_records(),
  demonstrate_binary(),
  demonstrate_processes(),
  demonstrate_ets(),

  io:format("~n--- Summary ---~n"),
  io:format("Erlang data structures:~n"),
  io:format("  - List: linked, head|tail, O(n) access~n"),
  io:format("  - Tuple: fixed-size, element() access~n"),
  io:format("  - Map: key-value, hash-based~n"),
  io:format("  - Set: unique elements~n"),
  io:format("  - Record: named tuples with field access~n"),
  io:format("  - Binary: packed binary data~n"),
  io:format("  - Process: lightweight actor model~n"),
  io:format("  - ETS: in-memory term storage~n"),
  io:format("  - Immutability: all data is immutable~n"),
  io:format("  - Pattern matching: primary control flow~n"),
  io:format("  - Let it crash: fault-tolerance philosophy~n").

File: examples/clojure/05-data-structures/src/data_structures.clj
; Data Structures in Clojure
; ==========================
; Clojure provides persistent data structures (lists, vectors, maps, sets).
; Data is immutable by default, making them safe for concurrent use.

(ns data-structures)

;; ── Vectors (indexed collections) ───────────────────────

(defn demonstrate-vectors []
  (println "--- Vectors (indexed, persistent) ---")

  ;; Creation with vector literal
  (def fruits ["apple" "banana" "cherry"])
  (println "fruits =" fruits)

  ;; Indexing (O(log n))
  (println "fruits[0] =" (get fruits 0))   ; apple
  (println "fruits[2] =" (get fruits 2))   ; cherry

  ;; nth is safer than get for out-of-range
  (println "nth 5 =" (nth fruits 5 "default"))  ; Returns "default"

  ;; conj adds to end of vector
  (def fruits-2 (conj fruits "date"))
  (println "After conj 'date':" fruits-2)

  ;; assoc updates index
  (def fruits-3 (assoc fruits 1 "coconut"))
  (println "After assoc index 1:" fruits-3)

  ;; subvec (slice)
  (println "fruits[1:3] =" (subvec fruits 1 3))

  ;; Range to create vector
  (def squares (vec (map #(* % %) (range 1 11)))
  (println "squares (1-10):" squares)

  ;; count (length)
  (println "count =" (count fruits)))

;; ── Lists (linked lists) ─────────────────────────────

(defn demonstrate-lists []
  (println "\n--- Lists (linked, persistent) ---")

  ;; Creation with list literal
  (def items '(1 2 3 4 5))
  (println "items =" items)

  ;; cons adds to beginning (prepend)
  (def items-2 (conj items 0))
  (println "After conj 0:" items-2)

  ;; rest and first (head and tail)
  (println "first =" (first items))   ; 1
  (println "rest =" (rest items))     ; (2 3 4 5)

  ;; lazy sequences
  (def naturals (iterate inc 1))
  (println "First 5 naturals:" (take 5 naturals))

  ;; Persistent! collections return new instances
  (def items-3 items)
  (println "Original list:" items)
  (println "Conj returns new list:" items-2)
  (println "Original unchanged:" items))

;; ── Maps (hash maps) ─────────────────────────────

(defn demonstrate-maps []
  (println "\n--- Maps (key-value) ---")

  ;; Creation with hash-map literal
  (def person {:name "Ada" :age 36 :city "London"})
  (println "person =" person)

  ;; Access with keywords (preferred)
  (println "name =" (:name person))
  (println "age =" (:age person))

  ;; Access with keys (also works)
  (println "name via key =" (get person :name))
  (println "age via key =" (get person :age))

  ;; assoc adds/updates key
  (def person-2 (assoc person :email "ada@example.com"))
  (println "After assoc email:" person-2)

  ;; dissoc removes key
  (def person-3 (dissoc person :city))
  (println "After dissoc city:" person-3)

  ;; contains?
  (println "contains :email?" (contains? person :email))
  (println "contains :phone?" (contains? person :phone))

  ;; keys and vals
  (println "keys:" (keys person))
  (println "vals:" (vals person))

  ;; sorted map (tree-based, maintains order)
  (def ordered-scores (sorted-map-by :score [{:name "Bob" :score 95} {:name "Alice" :score 87}]))
  (println "Sorted map (by score):" ordered-scores)

  ;; update (merge)
  (def person-4 (merge person {:country "UK"}))
  (println "After merge:" person-4))

;; ── Sets (unique collections) ─────────────────────

(defn demonstrate-sets []
  (println "\n--- Sets (unique elements) ---")

  ;; Creation with hash-set literal (duplicates removed)
  (def tags #{"python" "typing" "python" "data"})
  (println "tags =" tags)
  (println "count =" (count tags))  ; 3

  ;; conj adds element
  (def tags-2 (conj tags "algorithms"))
  (println "After conj 'algorithms':" tags-2)

  ;; disj removes element
  (def tags-3 (disj tags "data"))
  (println "After disj 'data':" tags-3)

  ;; contains?
  (println "contains 'typing'?" (contains? tags "typing"))

  ;; Set operations
  (def set-a #{1 2 3 4 5})
  (def set-b #{4 5 6 7 8})

  (println "union:" (clojure.set/union set-a set-b))
  (println "intersection:" (clojure.set/intersection set-a set-b))
  (println "difference a-b:" (clojure.set/difference set-a set-b))

  ;; sorted set (tree-based, maintains order)
  (def numbers (sorted-set #{5 2 8 1 9 3 7 6 4}))
  (println "Sorted set:" numbers)  ; Already sorted!

  ;; Subset/superset
  (println "subset? (2 3 4) of (1 2 3 4 5 6)?" (clojure.set/subset? #{2 3 4} #{1 2 3 4 5 6}))
  (println "superset? (1 2 3 4 5 6) of (2 3 4)?" (clojure.set/superset? #{1 2 3 4 5 6} #{2 3 4}))

;; ── Persistent! Immutability ─────────────────────────

(defn demonstrate-persistence []
  (println "\n--- Persistent! Immutability ---")

  (def original [1 2 3])
  (def modified (conj original 4))

  (println "Original:" original)
  (println "Modified (new vector):" modified)
  (println "Original unchanged:" original))

  ;; Transient for batch operations
  (println "Transient for batch operations:")
  (def transient-original [1 2 3])
  (def transient-result
    (persistent!  (reduce conj! transient-original [4 5])))
  (println "Result:" transient-result))

;; ── StructMaps (typed records) ─────────────────────

(defn demonstrate-structmaps []
  (println "\n--- StructMaps (typed records) ---")

  ;; Define struct type
  (defrecord Point [x y])

  ;; Creation
  (def p (->Point 10 20))
  (println "Point:" p)

  ;; Access fields
  (println "x =" (:x p))
  (println "y =" (:y p))

  ;; Assoc creates new instance
  (def p-2 (assoc p :x 30))
  (println "After assoc x:" p-2)
  (println "Original unchanged:" p)

  ;; Update multiple fields
  (def p-3 (assoc p :x 30 :y 40))
  (println "After update x and y:" p-3)

;; ── Zippers (structural navigation) ─────────────────

(defn demonstrate-zippers []
  (println "\n--- Zippers (structural navigation) ---")

  (def data [1 2 3 4 5])

  ;; Create zipper (focus point)
  (def zipped (vector-zipper data))

  (println "Current:" (zipped zipped))
  (println "Right:" (next zipped))
  (println "Left:" (prev zipped))

  ;; Move through structure
  (def zipped-2 (-> zipped next))
  (println "After next:" (zipped-2 zipped))

  ;; Replace and rebuild
  (def rebuilt (zipper/replace zipped-2 0 99))
  (def result (zipper/root rebuilt))
  (println "Rebuilt vector:" result)

;; ── Lazy sequences ─────────────────────────────────

(defn demonstrate-lazy []
  (println "\n--- Lazy sequences ---")

  ;; range creates lazy sequence
  (def lazy-naturals (range))
  (println "First 5 of lazy:" (take 5 lazy-naturals))

  ;; lazy cat
  (def combined (lazy-cat (repeat [1 2 3]) (repeat [4 5])))
  (println "First 10 of combined:" (take 10 combined))

  ;; lazy map/filter
  (def lazy-squares (map #(* % %) (range 1 100)))
  (println "First 5 lazy squares:" (take 5 lazy-squares))

  (println "Lazy sequences are only computed when needed!")

;; ── Summary ───────────────────────────────────────────

(defn -main []
  (println "Data Structures in Clojure")
  (println "==========================")

  demonstrate-vectors)
  demonstrate-lists)
  demonstrate-maps)
  demonstrate-sets)
  demonstrate-persistence)
  demonstrate-structmaps)
  demonstrate-zippers)
  demonstrate-lazy)

  (println "\n--- Summary ---")
  (println "Clojure data structures:")
  (println "  - Vector: indexed, persistent, O(log n) access")
  (println "  - List: linked, persistent, head/tail operations")
  (println "  - Map: hash map (unordered), sorted map (ordered)")
  (println "  - Set: hash set, sorted set")
  (println "  - Persistent!: immutable by default, structural sharing")
  (println "  - Transient: mutable for batch operations")
  (println "  - StructMaps: typed records with field access")
  (println "  - Zippers: structural navigation and modification")
  (println "  - Lazy sequences: range, lazy-cat, computed on demand")
  (println "  - Concurrency safe: immutable data prevents data races"))
